mux.js

```diff
@@ -0,0 +1,248 @@
+var protocol = require('hypercore-protocol')
+var readify = require('./ready')
+var inherits = require('inherits')
+var events = require('events')
+var debug = require('debug')('multifeed')
+var hypercore = require('hypercore')
+var xtend = require('xtend')
+
+// constants
+var MULTIFEED = 'MULTIFEED'
+var PROTOCOL_VERSION = '2.0.0'
+// extensions
+var MANIFEST = 'MANIFEST'
+var REQUEST_FEEDS = 'REQUEST_FEEDS'
+/*
+var ANNOUNCE_FEED = 'ANNOUNCE_FEED'
+var REQUEST_FEED_SIGNATURE = 'REQUEST_FEED_SIGNATURE'
+var FEED_SIGNATURE = 'FEED_SIGNATURE'
+*/
+
+var SupportedExtensions = [
+  MANIFEST,
+  REQUEST_FEEDS
+  //ANNOUNCE_FEED
+  //REQUEST_MANIFEST,
+  //REQUEST_FEED_SIGNATURE,
+  //FEED_SIGNATURE,
+]
+
+// `key` - protocol encryption key
+// `opts`- hypercore-protocol opts
+function Multiplexer (key, opts) {
+  if (!(this instanceof Multiplexer)) return new Multiplexer(key, opts)
+  debug('[REPLICATION] New mux initialized', key.toString('hex'), opts)
+  var self = this
+  self._opts = opts = opts || {}
+  self.extensions = opts.extensions = SupportedExtensions || opts.extensions
+
+  // initialize
+  self._localHave = null
+  self._localWant = null
+  self._remoteHas = null
+  self._remoteWants = null
+
+  var stream = this.stream = protocol(Object.assign(opts,{
+    userData: Buffer.from(JSON.stringify({
+      client: MULTIFEED,
+      version: PROTOCOL_VERSION,
+      // Help! exposing available extensions might be good for future version tolerance,
+      // but at the same time we're poentiallyleaking our user-agent fingerprint to a third party.
+      extensions: self.extensions
```

Comment:
It might not be needed -- right now we can infer the extensions available using the multifeed protocol version.